tls/s2n_x509_validator.c

```diff
@@ -549,3 +555,71 @@ s2n_cert_validation_code s2n_x509_validator_validate_cert_stapled_ocsp_response(
     return ret_val;
 #endif /* S2N_OCSP_STAPLING_SUPPORTED */
 }
+
+DEFINE_POINTER_CLEANUP_FUNC(STACK_OF(X509)*, wipe_cert_chain);
+
+S2N_RESULT s2n_x509_validator_validate_certificate_signatures(struct s2n_connection *conn, struct s2n_x509_validator *validator,
+                                                                             s2n_cert_validation_code *validation_code)
+{
+    ENSURE_REF(conn);
+    ENSURE_REF(validator);
+
+    if (conn->config->security_policy->certificate_signature_preferences == NULL) {
+        *validation_code = S2N_CERT_OK;
+        return S2N_RESULT_OK;
+    }
+    
+    DEFER_CLEANUP(STACK_OF(X509) *validated_chain = X509_STORE_CTX_get1_chain(validator->store_ctx), wipe_cert_chain_pointer);
+
+    const int certs_in_chain = sk_X509_num(validated_chain);
+
+    /* Do not validate the root certificate */
+    unsigned int certs_to_validate = certs_in_chain - 1;
+    
+    for (size_t i = 0; i < certs_to_validate; i++) {
+        bool out = false;
+        X509 *cert = sk_X509_value(validated_chain, i);
+
+        GUARD_RESULT(s2n_is_certificate_sig_scheme_supported(conn, cert, conn->config->security_policy->certificate_signature_preferences, &out));
+        if(out == false) {
+            *validation_code = S2N_CERT_ERR_UNTRUSTED;
+            return S2N_RESULT_OK;
+        }
+    }
+
+    *validation_code = S2N_CERT_OK;
```

Comment:
Why do we need this validation code? Does it provide any additional information? Won't all errors mean S2N_CERT_ERR_UNTRUSTED?

And what if we bail early on one of the GUARDs?